py/desispec/scripts/proc.py

```diff
@@ -722,26 +731,61 @@ def main(args=None, comm=None):
+                #- On Perlmutter GPU, --extract-size 22 provides the best performance.
+                #- 22 ranks = 5 ranks per GPU * 4 GPUs + 2 ranks for IO
+                extract_size = args.extract_size
+                if extract_size is None:
+                    import cupy as cp
+                    ngpus = cp.cuda.runtime.getDeviceCount()
+                    if rank == 0:
+                        log.info(f"{rank} found {ngpus} gpus")
+                    extract_ranks_per_gpu = 1
+                    extract_ranks_io = 2
+                    extract_size = extract_ranks_io + ngpus * extract_ranks_per_gpu
+                extract_group = 0
+                num_extract_groups = 1
+                extract_ranks = list(range(extract_size))
+                #- Create a comm group for the ranks that will perform extraction
+                if rank in extract_ranks:
+                    extract_incl = comm.group.Incl(extract_ranks)
+                    comm_extract = comm.Create_group(extract_incl)
+                    from gpu_specter.mpi import ParallelIOCoordinator
+                    coordinator = ParallelIOCoordinator(comm_extract)
+                else:
+                    run_extraction=False
+            else:
+                #- split communicator by 20 (number of bundles)
+                extract_size = 20
+                if (rank == 0) and (size%extract_size != 0):
+                    log.warning('MPI size={} should be evenly divisible by {}'.format(
+                        size, extract_size))
 
-            extract_group = rank // extract_size
-            num_extract_groups = (size + extract_size - 1) // extract_size
-            comm_extract = comm.Split(color=extract_group)
+                extract_group = rank // extract_size
+                num_extract_groups = (size + extract_size - 1) // extract_size
+                comm_extract = comm.Split(color=extract_group)
 
             for i in range(extract_group, len(args.cameras), num_extract_groups):
                 camera = args.cameras[i]
-                if camera in cmds:
+                if camera in cmds and run_extraction:
                     cmdargs = cmds[camera].split()[1:]
                     extract_args = desispec.scripts.extract.parse(cmdargs)
 
                     if comm_extract.rank == 0:
                         print('RUNNING: {}'.format(cmds[camera]))
 
-                    desispec.scripts.extract.main_mpi(extract_args, comm=comm_extract)
+                    if args.gpuextract:
+                        desispec.scripts.extract.main_gpu_specter(extract_args, coordinator=coordinator)
+                    elif args.gpuspecter:
+                        desispec.scripts.extract.main_gpu_specter(extract_args, comm=comm_extract)
+                    else:
+                        desispec.scripts.extract.main_mpi(extract_args, comm=comm_extract)
+                    comm_extract.barrier()
```

Comment:
Since this barrier is inside a loop and inside an if statement, double check the logic to make sure that no ranks could miss the barrier under some special case (e.g. if there are more ranks provided than needed for the number of cameras)

Reply:
I suspect this barrier will prevent the parallel IO coordinator from doing it's magic

Reply:
It worked for 32 ranks and extract_size=22 (so 10 idle cameras), but I agree it's better not to have it generally. I  removed the barrier and moved the check for missing files that it was introduced for in the first place to after the global barrier / serial extraction.